rosbag2_transport/test/rosbag2_transport/test_play_for.cpp

```diff
@@ -0,0 +1,275 @@
+
+class RosBag2PlayForTestFixture : public RosBag2PlayTestFixture
+{
+public:
+  static constexpr const char * kTopic1Name{"topic1"};
+  static constexpr const char * kTopic2Name{"topic2"};
+  static constexpr const char * kTopic1{"/topic1"};
+  static constexpr const char * kTopic2{"/topic2"};
+
+  std::vector<rosbag2_storage::TopicMetadata> get_topic_types()
+  {
+    return {{kTopic1Name, "test_msgs/BasicTypes", "", ""},
+      {kTopic2Name, "test_msgs/Arrays", "", ""}};
+  }
+
+  std::vector<std::shared_ptr<rosbag2_storage::SerializedBagMessage>>
+  get_serialized_messages()
+  {
+    auto primitive_message1 = get_messages_basic_types()[0];
+    primitive_message1->int32_value = kIntValue;
+
+    auto complex_message1 = get_messages_arrays()[0];
+    complex_message1->float32_values = {{kFloat1Value, kFloat2Value, kFloat3Value}};
+    complex_message1->bool_values = {{kBool1Value, kBool2Value, kBool3Value}};
+
+    // @{ Ordering matters. The mock reader implementation moves messages
+    //    around without any knowledge about message chronology. It just picks
+    //    the next one Make sure to keep the list in order or sort it!
+    std::vector<std::shared_ptr<rosbag2_storage::SerializedBagMessage>> messages =
+    {serialize_test_message(kTopic1Name, 500, primitive_message1),
+      serialize_test_message(kTopic2Name, 550, complex_message1),
+      serialize_test_message(kTopic1Name, 700, primitive_message1),
+      serialize_test_message(kTopic2Name, 750, complex_message1),
+      serialize_test_message(kTopic1Name, 900, primitive_message1),
+      serialize_test_message(kTopic2Name, 950, complex_message1)};
+    // @}
+    return messages;
+  }
+
+  void SetUp() override
+  {
+    auto topic_types = get_topic_types();
+    auto messages = get_serialized_messages();
+
+    auto prepared_mock_reader = std::make_unique<MockSequentialReader>();
+    prepared_mock_reader->prepare(messages, topic_types);
+    auto reader = std::make_unique<rosbag2_cpp::Reader>(std::move(prepared_mock_reader));
+
+    // Due to a problem related to the subscriber, we play many (3) messages but make the subscriber
+    // node spin only until 2 have arrived. Hence the 2 as `launch_subscriber()` argument.
+    sub_->add_subscription<test_msgs::msg::BasicTypes>(kTopic1, 2);
+    sub_->add_subscription<test_msgs::msg::Arrays>(kTopic2, 2);
+
+    player_ = std::make_shared<rosbag2_transport::Player>(
+      std::move(
+        reader), storage_options_, play_options_);
+  }
+
+  std::shared_ptr<rosbag2_transport::Player> player_;
+  std::future<void> await_received_messages_;
+};
+
+TEST_F(RosBag2PlayForTestFixture, play_for_all_are_played_due_to_duration)
+{
+  auto await_received_messages = sub_->spin_subscriptions();
+
+  player_->play(std::chrono::nanoseconds(std::chrono::milliseconds(1000)).count());
+
+  await_received_messages.get();
+
+  auto replayed_test_primitives = sub_->get_received_messages<test_msgs::msg::BasicTypes>(
+    kTopic1);
```

Comment:
@gbiggs nit: 
```suggestion
  auto replayed_test_primitives = 
      sub_->get_received_messages<test_msgs::msg::BasicTypes>(kTopic1);
```
Same elsewhere.